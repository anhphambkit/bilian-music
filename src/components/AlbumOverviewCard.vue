<template>
	<div class="album-overview-card">
		<b-skeleton-wrapper
			:loading="skeletonMixin_loading"
			class="skeleton-loading album-overview-card-loading"
		>
			<template #loading>
				<div class="skeleton-loading__partial skeleton-loading__top album-overview-card__top">
					<b-skeleton-img></b-skeleton-img>
				</div>
				<div class="skeleton-loading__partial skeleton-loading__bot album-overview-card__bot">
					<b-skeleton width="85%" class="mb-2"></b-skeleton>
					<b-skeleton width="55%" class="mt-2 mb-2"></b-skeleton>
				</div>
			</template>
			<slot>
				<div
					@click="$_navigatorMixin_goToAlbumDetail(albumId)"
					class="album-overview-card__top"
				>
					<b-img
						rounded
						@error="$_imageMixins_loadImageError"
						:src="$_imageMixins_loadImageSrc(albumImage)"
						:alt="albumName"
						class="album-overview-card__image"
					/>
					<div class="album-overview-card__sub-wrapper">
						<div class="album-overview-card__sub-info track">
							<b-icon-music-note-list></b-icon-music-note-list>
							{{ trackCount }}
						</div>
						<div class="album-overview-card__sub-info">
							<b-icon-disc-fill></b-icon-disc-fill>
							{{ discCount }}
						</div>
					</div>
				</div>
				<div
					class="album-overview-card__bot"
				>
					<div class="album-overview-card__info name" @click="$_navigatorMixin_goToAlbumDetail(albumId)">
						{{ albumName }}
					</div>
					<div class="album-overview-card__info date-released">
						{{ released }}
					</div>
				</div>
			</slot>
		</b-skeleton-wrapper>
	</div>
</template>
<script>
import SkeletonMixin from "@/mixins/SkeletonMixin";
import NavigatorMixins from "@/mixins/NavigatorMixins";
export default {
	name: "AlbumOverviewCard",
	mixins: [SkeletonMixin, NavigatorMixins],
	props: {
		discCount: {
			type: Number,
			default: 0,
		},
		trackCount: {
			type: Number,
			default: 0,
		},
		released: {
			type: String,
			required: true,
		},
		albumId: {
			type: String,
			required: true,
		},
		albumImage: {
			type: String,
			required: true,
		},
		albumName: {
			type: String,
			required: true,
		},
	},
};
</script>
<style lang="scss" scoped>
.album-overview-card {
	display: flex;
	flex-direction: column;
	padding: 10px;
	transition: all 0.3s ease;
	&:hover {
		box-shadow: 0px 2px 10px rgb(0, 0, 0);
	}
	&__top {
		position: relative;
	}
	&__image {
		width: 100%;
		height: auto;
		cursor: pointer;
	}
	&__sub-wrapper {
		position: absolute;
		top: 0;
		left: 0;
		padding: 5px 10px;
		background-color: #232323d4;
		display: flex;
		align-items: center;
		color: white;
	}
	&__sub-info {
		font-size: 12px;
		&.track {
			margin-right: 10px;
		}
	}
	&__bot {
		display: flex;
		flex-direction: column;
		margin-top: 10px;
	}
	&__info {
		&.date-released {
			color: #9c9ca2;
			font-size: 12px;
		}
		&.name {
			cursor: pointer;
		}
	}
}
</style>