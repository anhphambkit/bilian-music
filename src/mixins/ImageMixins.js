import DEFAULT_THUMB from "@/assets/media/default.jpg";
export default {
    name: 'ImageMixins',
    mixins: [],
    props: {},
    data() {
      return {}
    },
    computed: {},
    watch: {},
    created() {},
    mounted() {},
    methods: {
        $_imageMixins_loadImageSrc(src, altImage = DEFAULT_THUMB) {
            return src ? src : altImage;
        },
        $_imageMixins_loadImageError(event) {
            let img = event.srcElement;
            img.src = DEFAULT_THUMB;
            event.target.src = DEFAULT_THUMB;
            img.onerror = null; 
        }
    },
}