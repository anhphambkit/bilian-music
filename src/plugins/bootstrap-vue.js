import Vue from "vue";
// Import Bootstrap and BootstrapVue CSS files (order is important)
import 'bootstrap/dist/css/bootstrap.css'
import 'bootstrap-vue/dist/bootstrap-vue.css'

import {
	BContainer,
	BImg,
	BImgLazy,
	BRow,
	BCol,
	BBadge,
	TooltipPlugin,
	BIconMusicNoteList,
	BIconSkipStartFill,
	BIconSkipEndFill,
	BIconDiscFill,
	BSkeletonWrapper,
	BSkeleton,
	BSkeletonImg,
	BOverlay
} from "bootstrap-vue";

Vue.component("b-container", BContainer);
Vue.component("b-img", BImg);
Vue.component("b-img-lazy", BImgLazy);
Vue.component("b-row", BRow);
Vue.component("b-col", BCol);
Vue.component("b-badge", BBadge);
Vue.component("b-icon-music-note-list", BIconMusicNoteList);
Vue.component("b-icon-skip-start-fill", BIconSkipStartFill);
Vue.component("b-icon-skip-end-fill", BIconSkipEndFill);
Vue.component("b-icon-disc-fill", BIconDiscFill);
Vue.component("b-skeleton-wrapper", BSkeletonWrapper);
Vue.component("b-skeleton", BSkeleton);
Vue.component("b-skeleton-img", BSkeletonImg);
Vue.component("b-overlay", BOverlay);
Vue.use(TooltipPlugin);