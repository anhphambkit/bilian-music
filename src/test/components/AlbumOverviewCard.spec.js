import { createLocalVue, shallowMount } from '@vue/test-utils'
import AlbumOverviewCard from '@/components/AlbumOverviewCard'
import ImageMixins from "@/mixins/ImageMixins";

// install plugins as normal
import { BootstrapVueIcons, BootstrapVue } from 'bootstrap-vue';
function createConfig(overrides) {
    process.env.BOOTSTRAP_VUE_NO_WARN = true
    const albumItem = { 
        released: "2020-10-23T00:00:00.000Z",
        albumId: "alb.517280316",
        albumImage: "https://api.napster.com/imageserver/v2/albums/alb.517280316/images/200x200.jpg",
        albumName: "Year Zero",
        artistImage: "https://api.napster.com/imageserver/v2/artists/art.122743427/images/70x70.jpg",
        artistName: "Tchami",
        artistId: "art.122743427",
    }
    const localVue = createLocalVue()
    localVue.mixin(ImageMixins)
    localVue.use(BootstrapVue)
    localVue.use(BootstrapVueIcons)
    const mocks = {
        $router: {
          push: jest.fn()
        },
    }
    const propsData = albumItem
    return Object.assign({ mocks, propsData, localVue }, overrides)
}
  
describe('AlbumOverviewCard.vue', () => {
    test('TEST CASE 1: Render AlbumOverViewCard HTML', () => {
        const wrapper  = shallowMount(AlbumOverviewCard, createConfig())
        expect(wrapper).toMatchSnapshot()
    })

    test('TEST CASE 2: redirect to AlbumDetail page when clicking on image album', () => {
        const config = createConfig()
        const wrapper  = shallowMount(AlbumOverviewCard, config)
        wrapper.find('.album-overview-card__top').trigger('click')
        const spy = jest.spyOn(config.mocks.$router, 'push')
        expect(spy).toHaveBeenCalledWith(`/album/${config.propsData.albumId}`)
    })

    test('TEST CASE 3: redirect to AlbumDetail page when clicking on album name', () => {
        const config = createConfig()
        const wrapper  = shallowMount(AlbumOverviewCard, config)
        wrapper.find('.album-overview-card__info.name').trigger('click')
        const spy = jest.spyOn(config.mocks.$router, 'push')
        expect(spy).toHaveBeenCalledWith(`/album/${config.propsData.albumId}`)
    })
})