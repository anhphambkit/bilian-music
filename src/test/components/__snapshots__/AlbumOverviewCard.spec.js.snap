// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumOverviewCard.vue TEST CASE 1: Render AlbumOverViewCard HTML 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
